go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -639,32 +672,6 @@ func checkAndRecoverDeadPrimary(analysisEntry inst.ReplicationAnalysis, candidat
 	}
 	log.Infof("Analysis: %v, deadprimary %+v", analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey)
 
-	// this check is needed because sometimes DeadPrimary code path is forcefully spawned off from other recoveries like PrimaryHasPrimary.
-	// So we need to check that we only run an ERS if the instance that we analyzed was actually a primary! Otherwise, we would end up running an ERS
-	// even when the cluster is fine or the problem can be fixed via some other recovery
-	if tablet.Type != topodatapb.TabletType_PRIMARY {
-		RefreshTablets(true /* forceRefresh */)
-		AuditTopologyRecovery(topologyRecovery, "another agent seems to have fixed the problem")
-		return false, topologyRecovery, nil
-	}
-
-	// check if we have received SIGTERM, if we have, we should not continue with the recovery
-	val := atomic.LoadInt32(&hasReceivedSIGTERM)
-	if val > 0 {
-		return false, topologyRecovery, errors.New("Can't lock shard: SIGTERM received")
-	}
-
-	// check if we have received an ERS in progress, if we do, we should not continue with the recovery
-	if checkAndSetIfERSInProgress() {
-		AuditTopologyRecovery(topologyRecovery, "an ERS is already in progress, not issuing another")
-		return false, topologyRecovery, nil
-	}
-	defer setERSCompleted()
-
```

Comment:
Is it safe to remove this?

Reply:
Yes it safe to remove now. Let me start with why it was introduced in the first place. I introduced this logic when I initially changed the `checkAndRecoverDeadPrimary` to use `EmergencyReparentShard` in https://github.com/vitessio/vitess/pull/8492.
Before this PR, we used to lock the shard as part of the `ERS` code, i.e. after we have called `ERS`. Because the cluster failure checks happen on time ticks and repair operations in separate go routines, we sometimes had situations where two ticks of failure check spawned two different threads of repair, each which would have tried to run an `ERS`.

The `LockShard` functionality is blocking and doesn't return an error if the shard is locked until the timeout. This meant that if there were two calls to `ERS`, one would get the lock and the other would wait for it. After the first finished its execution, we had another run of `ERS`. So, we had two runs of `ERS` for a single failure, which was not correct. The solution that we decided on then, was to add this code, which was a mutex lock and a variable to know if ERS was in progress from this same VTOrc instance. This prevented us from spawning another `ERS` call when one was already running.

This was a temporary workaround till we addressed federation like we have now. Since we lock the shard much earlier on and check again if the failure is required after refreshing the information, we no longer need this type of locking. Even if there is a second call to `checkAndRecoverDeadPrimary` it would block on acquiring the lock in the beginning, and once it does, it will reload the information and then decide whether to run `ERS` or not. So, if the first call to `ERS` fixed everything, there would be nothing to do and we exit out.

So, the change in this PR actually is powerful enough to address three concerns - 
1. Redundant operations from the same VTOrc instance (Like we see above)
2. Redundant operations from VTOrc when a cluster operation is run from `vtctld` (Like PRS, which I have added a test for now)
3. Federation between different VTOrc instances. With this change, it should be safe to have multiple VTOrc instances without them stepping on each others toes because this change guarantees us that we only run a cluster operation when it is needed.

For the sake of completeness, I will also add that this reload of information has been added to 3 of the 5 functions we use for fixing things. The two that I have not yet added this to, are `fixPrimary` and `fixReplica` which only fix the replication parameters on the primary and replica respectively. I intend to make that change as well, but I didn't want to add any code that didn't have an associated end to end test with it. So, I'll make this change in the next PR which first refactors the way we get the replication analysis and run the operations. Which will enable us to add unit tests for this change too.